### PR TITLE
New Resource: aws_wafregional_rate_based_rule

### PR DESCRIPTION
**Pull request by radeksimko**
## Test results

```
make testacc TEST=./aws TESTARGS='-run=TestAccAWSWafRegionalRateBasedRule_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSWafRegionalRateBasedRule_ -timeout 120m
=== RUN   TestAccAWSWafRegionalRateBasedRule_basic
--- PASS: TestAccAWSWafRegionalRateBasedRule_basic (48.24s)
=== RUN   TestAccAWSWafRegionalRateBasedRule_changeNameForceNew
--- PASS: TestAccAWSWafRegionalRateBasedRule_changeNameForceNew (91.79s)
=== RUN   TestAccAWSWafRegionalRateBasedRule_disappears
--- PASS: TestAccAWSWafRegionalRateBasedRule_disappears (47.80s)
=== RUN   TestAccAWSWafRegionalRateBasedRule_changePredicates
--- PASS: TestAccAWSWafRegionalRateBasedRule_changePredicates (77.09s)
=== RUN   TestAccAWSWafRegionalRateBasedRule_noPredicates
--- PASS: TestAccAWSWafRegionalRateBasedRule_noPredicates (28.84s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	293.792s
```

### CODE REVIEW

File: aws/resource_aws_wafregional_rate_based_rule.go [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+			return nil
+		}
+
+		return err
+	}
+
+	var predicates []map[string]interface{}
+
+	for _, predicateSet := range resp.Rule.MatchPredicates {
+		predicate := map[string]interface{}{
```
Comment (appilon) [inline aws/resource_aws_wafregional_rate_based_rule.go]: I don't know if I can argue it's better but you could avoid the `predicate` variable
```
predicates = append(predicates, map[string]interface{}{
  "negated": *predicateSet.Negated,
  "type":    *predicateSet.Type,
  "data_id": *predicateSet.DataId,
})
```

File: aws/resource_aws_wafregional_rate_based_rule.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+	return nil
+}
+
+func updateWafRateBasedRuleResourceWR(id string, oldP, newP []interface{}, rateLimit interface{}, conn *wafregional.WAFRegional, region string) error {
+	wr := newWafRegionalRetryer(conn, region)
+	_, err := wr.RetryWithToken(func(token *string) (interface{}, error) {
+		req := &waf.UpdateRateBasedRuleInput{
+			ChangeToken: token,
+			RuleId:      aws.String(id),
+			Updates:     diffWafRulePredicates(oldP, newP),
```
Comment (appilon) [inline aws/resource_aws_wafregional_rate_based_rule.go]: I suggest maybe moving `diffWafRulePredicates`'s definition to `waf_helpers.go`

File: aws/resource_aws_wafregional_rate_based_rule_test.go [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+
+func TestAccAWSWafRegionalRateBasedRule_changeNameForceNew(t *testing.T) {
+	var before, after waf.RateBasedRule
+	wafRuleName := fmt.Sprintf("wafrule%s", acctest.RandString(5))
+	wafRuleNewName := fmt.Sprintf("wafrulenew%s", acctest.RandString(5))
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSWafIPSetDestroy,
```
Comment (appilon) [inline aws/resource_aws_wafregional_rate_based_rule_test.go]: wrong destroy check?

File: aws/resource_aws_wafregional_rate_based_rule_test.go [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+	var byteMatchSet waf.ByteMatchSet
+
+	var before, after waf.RateBasedRule
+	var idx int
+	ruleName := fmt.Sprintf("wafrule%s", acctest.RandString(5))
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSWafRuleDestroy,
```
Comment (appilon) [inline aws/resource_aws_wafregional_rate_based_rule_test.go]: wrong destroy check?

File: aws/resource_aws_wafregional_rate_based_rule.go (hunk 1 above)
Comment (radeksimko) [reply aws/resource_aws_wafregional_rate_based_rule.go]: That's a valid argument 👍 I will address that.

File: aws/resource_aws_wafregional_rate_based_rule_test.go (hunk 3 above)
Comment (radeksimko) [reply aws/resource_aws_wafregional_rate_based_rule_test.go]: Good spot! 🙈

File: aws/resource_aws_wafregional_rate_based_rule_test.go (hunk 4 above)
Comment (radeksimko) [reply aws/resource_aws_wafregional_rate_based_rule_test.go]: Good spot! 🙈

**Comment by radeksimko** [comment]
@appilon PTAL, I believe I addressed all your feedback.

**Comment by bflad** [comment]
This has been released in [version 1.13.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1130-march-28-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**